src/library_webgpu.js

```diff
@@ -1212,60 +1258,133 @@ var LibraryWebGPU = {
-      dynCall('viiji', callback, [WEBGPU_BUFFER_MAP_ASYNC_STATUS_SUCCESS, data, {{{ sendI64Argument('dataLength_low', 'dataLength_high') }}}, userdata]);
-    }, function() {
-      // TODO(kainino0x): Figure out how to pick other error status values.
-      var WEBGPU_BUFFER_MAP_ASYNC_STATUS_ERROR = 1;
-      dynCall('viiji', callback, [WEBGPU_BUFFER_MAP_ASYNC_STATUS_ERROR, 0, 0, 0, userdata]);
+  wgpuBufferGetConstMappedRange: function(bufferId, offset, size) {
+    var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
+
+    var mapped;
+    try {
+      mapped = bufferWrapper.object["getMappedRange"](offset, size);
+    } catch (ex) {
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var data = _malloc(mapped.byteLength);
+    HEAPU8.set(new Uint8Array(mapped), data);
+    bufferWrapper.onUnmap.push(function() {
+      _free(data);
     });
+    return data;
   },
 
-  wgpuBufferMapWriteAsync: function(bufferId, callback, userdata) {
+  wgpuBufferGetMappedRange: function(bufferId, offset, size) {
     var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
+
+    if (bufferWrapper.mapMode !== 2 /* WGPUMapMode_Write */) {
+#if ASSERTIONS
+      abort("GetMappedRange called, but buffer not mapped for writing");
+#endif
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var mapped;
+    try {
+      mapped = bufferWrapper.object["getMappedRange"](offset, size);
+    } catch (ex) {
+      // TODO(kainino0x): Somehow inject a validation error?
+      return 0;
+    }
+
+    var data = _malloc(mapped.byteLength);
+    HEAPU8.fill(0, data, mapped.byteLength);
+    bufferWrapper.onUnmap.push(function() {
+      new Uint8Array(mapped).set(HEAPU8.subarray(data, data + mapped.byteLength));
+      _free(data);
+    });
+    return data;
+  },
+
+  wgpuBufferMapAsync: function(bufferId, mode, offset, size, callback, userdata) {
+    var bufferWrapper = WebGPU.mgrBuffer.objects[bufferId];
+    bufferWrapper.mapMode = mode;
+    bufferWrapper.onUnmap = [];
     var buffer = bufferWrapper.object;
 
-    var WEBGPU_BUFFER_MAP_ASYNC_STATUS_SUCCESS = 0;
-    var WEBGPU_BUFFER_MAP_ASYNC_STATUS_ERROR = 1;
-    buffer["mapWriteAsync"]().then(function(mapped) {
-      WebGPU.trackMapWrite(bufferWrapper, mapped);
+    // `callback` takes (WGPUBufferMapAsyncStatus status, void * userdata)
 
-      var data = bufferWrapper.mapWriteSrc;
-      var dataLength_high = (mapped.byteLength / 0x100000000) | 0;
-      var dataLength_low = mapped.byteLength | 0;
-      // WGPUBufferMapAsyncStatus status, void* data, uint64_t dataLength, void* userdata
-      dynCall('viiji', callback, [WEBGPU_BUFFER_MAP_ASYNC_STATUS_SUCCESS, data, {{{ sendI64Argument('dataLength_low', 'dataLength_high') }}}, userdata]);
+    buffer["mapAsync"](mode, offset, size).then(function() {
+      dynCall('vii', callback, [0 /* WGPUBufferMapAsyncStatus_Success */, userdata]);
```

Comment:
In all places when the function signature is statically known, we can directly call `dynCall_vii(func, param0, param1);` instead of `dynCall('vii', [param0, param1])`. The first form statically knows the signature it is calling, whereas the second form will need to dynamically bind the function call signature, so it is much slower. Also the second form will create a temp garbage array for the call.